File: app/src/main/java/pt/estga/spotme/ui/authentication/RegisterActivity.java
package pt.estga.spotme.ui.authentication;

import android.content.Intent;
import android.os.Bundle;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;
import androidx.appcompat.app.AppCompatActivity;
import pt.estga.spotme.MainActivity;
import pt.estga.spotme.R;
import pt.estga.spotme.database.AppDatabase;
import pt.estga.spotme.database.UserDao;
import pt.estga.spotme.entities.User;
import pt.estga.spotme.utils.UserSession;
import pt.estga.spotme.utils.PasswordUtils; // Classe para Hashing
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class RegisterActivity extends AppCompatActivity {

    private EditText usernameEditText, emailEditText, passwordEditText, confirmPasswordEditText, phoneEditText;
    private Button registerButton;
    private TextView loginTextView;
    private UserDao userDao;
    private ExecutorService executorService = Executors.newSingleThreadExecutor();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.register_view);

        // Inicializar BD
        AppDatabase db = AppDatabase.getInstance(this);
        userDao = db.userDao();

        // Inicializar os elementos da UI
        usernameEditText = findViewById(R.id.usernameEditText);
        emailEditText = findViewById(R.id.emailEditText);
        passwordEditText = findViewById(R.id.passwordEditText);
        phoneEditText = findViewById(R.id.phoneNumEditText);
        confirmPasswordEditText = findViewById(R.id.confirmPasswordEditText);
        registerButton = findViewById(R.id.registerButton);
        loginTextView = findViewById(R.id.loginText);

        // Ação do botão de registo
        registerButton.setOnClickListener(v -> registerUser());

        // Ir para a página de login
        loginTextView.setOnClickListener(v -> {
            startActivity(new Intent(RegisterActivity.this, LoginActivity.class));
            finish();
        });
    }

    private void registerUser() {
        String username = usernameEditText.getText().toString().trim();
        String email = emailEditText.getText().toString().trim();
        String phone = phoneEditText.getText().toString().trim();
        String password = passwordEditText.getText().toString().trim();
        String confirmPassword = confirmPasswordEditText.getText().toString().trim();

        if (username.isEmpty() || email.isEmpty() || password.isEmpty() || confirmPassword.isEmpty() || phone.isEmpty()) {
            showToast("Todos os campos devem ser preenchidos!");
            return;
        }

        if (!password.equals(confirmPassword)) {
            showToast("As senhas não coincidem!");
            return;
        }

        // Verifica se o utilizador já existe e regista-o
        executorService.execute(() -> {
            User existingUser = userDao.getUserByEmail(email);
            if (existingUser != null) {
                runOnUiThread(() -> showToast("Email já registado!"));
                return;
            }

            // Hash da senha antes de salvar
            String hashedPassword = PasswordUtils.hashPassword(password);

            // Criar e inserir novo usuário na BD
            User newUser = new User(username, hashedPassword, email, phone);
            userDao.insert(newUser);

            // Procurar o ID do novo utilizador
            User registeredUser = userDao.getUserByEmail(email);
            if (registeredUser != null) {
                UserSession session = UserSession.getInstance(getApplicationContext());
                session.setUser(registeredUser);
                session.setUserProfileImage(registeredUser.getProfileImage()); // Caso tenha imagem
                session.setUserPassword(password);

                runOnUiThread(() -> {
                    showToast("Conta registada com sucesso!");
                    startActivity(new Intent(RegisterActivity.this, MainActivity.class));
                    finish();
                });
            }
        });
    }

    private void showToast(String message) {
        runOnUiThread(() -> Toast.makeText(this, message, Toast.LENGTH_SHORT).show());
    }
}

File: app/src/main/java/pt/estga/spotme/ui/authentication/RegisterActivity.kt
package pt.estga.spotme.ui.authentication

import android.content.Intent
import android.os.Bundle
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.ViewModelProvider
import pt.estga.spotme.MainActivity
import pt.estga.spotme.database.AppDatabase
import pt.estga.spotme.database.repository.AuthRepository
import pt.estga.spotme.databinding.RegisterViewBinding
import pt.estga.spotme.entities.User
import pt.estga.spotme.utils.PasswordUtils
import pt.estga.spotme.utils.UserSession

class RegisterActivity : AppCompatActivity() {

    private lateinit var binding: RegisterViewBinding
    private lateinit var registerViewModel: RegisterViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = RegisterViewBinding.inflate(layoutInflater)
        setContentView(binding.root)

        val db = AppDatabase.getInstance(applicationContext)
        val factory = RegisterViewModelFactory(AuthRepository(db.userDao()))
        registerViewModel = ViewModelProvider(this, factory)[RegisterViewModel::class.java]

        binding.registerButton.setOnClickListener { handleRegister() }
        binding.loginText.setOnClickListener {
            startActivity(Intent(this, LoginActivity::class.java))
            finish()
        }

        observeRegistrationResult()
    }

    private fun handleRegister() {
        val username = binding.usernameEditText.text.toString().trim()
        val email = binding.emailEditText.text.toString().trim()
        val phone = binding.phoneNumEditText.text.toString().trim()
        val password = binding.passwordEditText.text.toString().trim()
        val confirmPassword = binding.confirmPasswordEditText.text.toString().trim()

        val validationError = AuthFormValidator.validateRegistration(
            username, email, phone, password, confirmPassword
        )

        if (validationError != null) {
            showToast(validationError)
            return
        }

        val hashedPassword = PasswordUtils.hashPassword(password)
        val newUser = User(
            username = username,
            email = email,
            phone = phone,
            password = hashedPassword,
            profileImage = ""
        )

        registerViewModel.register(newUser)
    }

    private fun observeRegistrationResult() {
        registerViewModel.registrationResult.observe(this) { result ->
            result.onSuccess { user ->
                UserSession.getInstance(applicationContext).updateUserSession(user, user.password)
                showToast("Conta registada com sucesso!")
                startActivity(Intent(this, MainActivity::class.java))
                finish()
            }
            result.onFailure { ex ->
                showToast(ex.message ?: "Erro desconhecido")
            }
        }
    }

    private fun showToast(message: String) {
        Toast.makeText(this, message, Toast.LENGTH_SHORT).show()
    }
}